File: src/TreeOrder.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x){
        val = x;
        left = null;
        right = null;
    }
}

public class TreeOrder {
    //递归前序遍历
    public void preOrder(TreeNode node) {
        if(node == null) {
            return;
        }
        System.out.print(node.val + " ");
        preOrder(node.left);
        preOrder(node.right);
    }
    //非递归前序遍历
    public void preOrder2(TreeNode node) {
        if (node == null) {
            return;
        }
        Deque<TreeNode> st = new LinkedList<>();
        st.push(node);
        while (st.size() != 0) {
            TreeNode tmp = st.pop();
            System.out.print(tmp.val + " ");
            if (tmp.right != null) {
                st.push(tmp.right);
            }
            if (tmp.left != null) {
                st.push(tmp.left);
            }
        }
    }

    //递归中序遍历
    public void inOrder(TreeNode node) {
        if(node == null) {
            return;
        }
        inOrder(node.left);
        System.out.print(node.val + " ");
        inOrder(node.right);
    }
    //非递归中序遍历
    public void inOrder2(TreeNode node) {
        if (node == null) {
            return;
        }
        Deque<TreeNode> st = new LinkedList<>();
        st.push(node);
        while (st.peek().left != null) {
            st.push(st.peek().left);
        }
        while (st.size() != 0) {
            TreeNode tmp = st.pop();
            System.out.print(tmp.val + " ");
            if (tmp.right != null) {
                st.push(tmp.right);
                while (st.peek().left != null) {
                    st.push(st.peek().left);
                }
            }
        }
    }

    //递归后序遍历
    public void postOrder(TreeNode node) {
        if(node == null) {
            return;
        }
        postOrder(node.left);
        postOrder(node.right);
        System.out.print(node.val + " ");
    }
    //非递归后序遍历
    public void postOrder2(TreeNode node) {
        if (node == null) {
            return;
        }
        Deque<TreeNode> st = new LinkedList<>();
        st.push(node);
        List<TreeNode> res = new ArrayList<>();
        while (st.size() != 0) {
            TreeNode tmp = st.pop();
            res.add(tmp);
            if (tmp.left != null) {
                st.push(tmp.left);
            }
            if (tmp.right != null) {
                st.push(tmp.right);
            }
        }
        Collections.reverse(res);
        for (int i = 0; i < res.size(); i++) {
            System.out.print(res.get(i).val + " ");
        }
    }

    //层序遍历
    public void levelOrder(TreeNode node) {
        if (node == null) {
            return;
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(node);
        while (q.size() != 0) {
            TreeNode tmp = q.poll();
            System.out.print(tmp.val + " ");
            if (tmp.left != null) {
                q.offer(tmp.left);
            }
            if (tmp.right != null) {
                q.offer(tmp.right);
            }
        }
    }

    public static void main(String[] args) {
        TreeOrder treeOrder = new TreeOrder();
        TreeNode root = new TreeNode(2);
        root.left = new TreeNode(7);
        root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);
        root.left.right.right = new TreeNode(11);
        root.right = new TreeNode(5);
        root.right.right = new TreeNode(9);
        root.right.right.left = new TreeNode(4);

        System.out.println("递归前序遍历：");
        treeOrder.preOrder(root);
        System.out.println();
        System.out.println("非递归前序遍历：");
        treeOrder.preOrder2(root);
        System.out.println();

        System.out.println("递归中序遍历：");
        treeOrder.inOrder(root);
        System.out.println();
        System.out.println("非递归中序遍历：");
        treeOrder.inOrder2(root);
        System.out.println();

        System.out.println("递归后序遍历：");
        treeOrder.postOrder(root);
        System.out.println();
        System.out.println("非递归后序遍历：");
        treeOrder.postOrder2(root);
        System.out.println();

        System.out.println("层序遍历：");
        treeOrder.levelOrder(root);
        System.out.println();
    }
}

File: py3/TreeOrder.py
from collections import deque
class TreeNode():
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class TreeOrder():
    def preOrder(self, root):
        if not root:
            return
        print(root.val, end=' ')
        self.preOrder(root.left)
        self.preOrder(root.right)
    def preOrder2(self, root):
        if not root:
            return
        st = [root]
        while st:
            tmp = st.pop()
            print(tmp.val, end=' ')
            if tmp.right:
                st.append(tmp.right)
            if tmp.left:
                st.append(tmp.left)
    def inOrder(self, root):
        if not root:
            return
        self.inOrder(root.left)
        print(root.val, end=' ')
        self.inOrder(root.right)
    def inOrder2(self, root):
        if not root:
            return
        st = [root]
        while st[-1].left:
            st.append(st[-1].left)
        while st:
            tmp = st.pop()
            print(tmp.val, end=' ')
            if tmp.right:
                st.append(tmp.right)
                while st[-1].left:
                    st.append(st[-1].left)
    def postOrder(self, root):
        if not root:
            return
        self.postOrder(root.left)
        self.postOrder(root.right)
        print(root.val, end=' ')
    def postOrder2(self, root):
        if not root:
            return
        st = [root]
        res = []
        while st:
            tmp = st.pop()
            res.append(tmp)
            if tmp.left:
                st.append(tmp.left)
            if tmp.right:
                st.append(tmp.right)
        for node in res[::-1]:
            print(node.val, end=' ')
    def levelOrder(self, root):
        if not root:
            return
        q = deque()
        q.append(root)
        while q:
            tmp = q.popleft()
            print(tmp.val, end=' ')
            if tmp.left:
                q.append(tmp.left)
            if tmp.right:
                q.append(tmp.right)
if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    treeOrder = TreeOrder()
    print('递归前序遍历：')
    treeOrder.preOrder(root)
    print('\n非递归前序遍历：')
    treeOrder.preOrder2(root)
    print('\n递归中序遍历：')
    treeOrder.inOrder(root)
    print('\n非递归中序遍历：')
    treeOrder.inOrder2(root)
    print('\n递归后序遍历：')
    treeOrder.postOrder(root)
    print('\n非递归后序遍历：')
    treeOrder.postOrder2(root)
    print('\n层次遍历：')
    treeOrder.levelOrder(root)